### PR TITLE
[OpenMP]Update use_device_clause lowering

### PR DESCRIPTION
**Pull request by TIFitis**
This patch updates the use_device_ptr and use_device_addr clauses to use the mapInfoOps for lowering. This allows all the types that are handle by the map clauses such as derived types to also be supported by the use_device_clauses.

This is patch 1/2 in a series of patches.

Co-authored-by: Raghu Maddhipatla raghu.maddhipatla@amd.com

### CODE REVIEW

File: flang/test/Lower/OpenMP/use-device-ptr-to-use-device-addr.f90
Comment (skatrak) [inline flang/test/Lower/OpenMP/use-device-ptr-to-use-device-addr.f90:8]: Nit: Lines in this comment seems to have been split very short. Since you're already making changes to it, might as well fix that too.
```suggestion
! This tests primary goal is to check the promotion of non-CPTR arguments from
! use_device_ptr to use_device_addr works, without breaking any functionality.
```

**Review by skatrak — COMMENTED** [review]
Thank you, these changes make sense to me. However, I see quite a bit of code duplication as a result, so I'd like to request addressing that so we avoid divergence problems in the future.

**Comment by TIFitis** [comment]
@skatrak Thanks for the review. I have addressed the comments in the latest revision.

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (skatrak) [inline flang/lib/Lower/OpenMP/OpenMP.cpp:897]: If this is only needed for `target data` and not `target`, I think we need a comment here to explain why. Otherwise, we should decide whether we are supposed to do this in both cases or in none of them, since it looks to me like something that may have slipped through the cracks due to having this code duplicated between `genBodyOfTargetOp` and `genBodyOfTargetDataOp`. Perhaps @agozillon can comment on that.

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (skatrak) [inline flang/lib/Lower/OpenMP/OpenMP.cpp:832]: In my opinion, this is not the best way to share map-related functionality between `target` and `target_data`. By making both call a single function and then conditionally doing things depending on who made the call, the function becomes harder to understand and maintain. My comment was in the direction of trying to simplify these `genBodyOfTargetOp` and `genBodyOfTargetDataOp` that had grown complex while having large similarities.

```c++
static void bindMapSymbols(...) {
  auto cloneBounds = [cloneBound](llvm::ArrayRef<mlir::Value> bounds) {
    llvm::SmallVector<mlir::Value> clonedBounds;
    for (mlir::Value bound : bounds)
      clonedBounds.emplace_back(cloneBound(bound));
    return clonedBounds;
  };
  // Bind the symbols to their corresponding block arguments.
  for (auto [argIndex, argSymbol] : llvm::enumerate(mapSyms)) {
    ...
  }
}

template<typename OpTy>
static void cloneBoundsDependencies(OpTy op, ...) {
  static_assert(std::is_same_v<OpTy, mlir::omp::TargetOp> ||
                std::is_same_v<OpTy, mlir::omp::TargetDataOp>,
                "OpTy must be TargetOp or TargetDataOp");
  // Check if cloning the bounds introduced any dependency on the outer region.
  // If so, then either clone them as well if they are MemoryEffectFree, or else
  // copy them to a new temporary and add them to the map and block_argument
  // lists and replace their uses with the new temporary.
  llvm::SetVector<mlir::Value> valuesDefinedAbove;
  mlir::getUsedValuesDefinedAbove(region, valuesDefinedAbove);
  while (!valuesDefinedAbove.empty()) {
    for (mlir::Value val : valuesDefinedAbove) {
      ...
      op.getMapVarsMutable().append(mapOp);
      ...
    }
  }
}

static void genBodyOfTargetDataOp(...) {
  ...
  firOpBuilder.createBlock(&region, {}, useDeviceTypes, useDeviceLocs);
  bindMapSymbols(...);
  cloneBoundsDependencies(dataOp, ...);
  ...
}

static void genBodyOfTargetOp(...) {
  ...
  llvm::SmallVector<mlir::Type> allRegionArgTypes;
  mergePrivateVarsInfo(...);
  ...
  auto *regionBlock = firOpBuilder.createBlock(&region, {}, allRegionArgTypes, allRegionArgLocs);
  bindMapSymbols(...);
  for (auto [argIndex, argSymbol] :
       llvm::enumerate(dsp.getAllSymbolsToPrivatize())) {
    ...
  }
  cloneBoundsDependencies(targetOp, ...);
  ...
}
```
Let me know what you think.

**Review by skatrak — COMMENTED** [review]
Thank you for working on my previous comments. I think the refactoring in `ClauseProcessor` is quite a nice change. I have some more comments to hopefully improve things a bit further before merging.

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (skatrak) [inline flang/lib/Lower/OpenMP/OpenMP.cpp:859]: Nit: There are several instances of this in this function. It's not necessary because `targetOps` is already of that type.
```suggestion
  mergePrivateVarsInfo(targetOp, mapSymTypes,
```

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (skatrak) [inline flang/lib/Lower/OpenMP/OpenMP.cpp:892]: Nit: Accidental change to the comment?

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (skatrak) [inline flang/lib/Lower/OpenMP/OpenMP.cpp:788]: Nit: In my opinion, this comment change doesn't seem like an improvement.

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (skatrak) [inline flang/lib/Lower/OpenMP/OpenMP.cpp:783]: Would it be correct for a bound cloned into the `omp.target_data` region to take an outside value as an argument? For `omp.target` this is disallowed due to the `IsolatedFromAbove` trait, so there's some additional post-processing to handle them. Here we don't have the same restriction, but I wonder if we would produce valid code if we don't map these too.

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (skatrak) [inline flang/lib/Lower/OpenMP/OpenMP.cpp:696]: Nit: Could this argument be removed and just access it from `region.front()`? This function is expected to be called while building the entry block, so we can just `assert(region.hasOneBlock())`.

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (skatrak) [inline flang/lib/Lower/OpenMP/OpenMP.cpp:801]: Nit: I think the comment was correct previously.

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (skatrak) [inline flang/lib/Lower/OpenMP/OpenMP.cpp:869]:
```suggestion
  mlir::Block *regionBlock = firOpBuilder.createBlock(&region, {}, allRegionArgTypes,
```

**Review by skatrak — COMMENTED** [review]
Thanks Akash again, this looks good. I have a couple of nits remaining and just one concern left.

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (TIFitis) [reply flang/lib/Lower/OpenMP/OpenMP.cpp:783]: Yes, it's okay for outside values to reach here.

**Comment by TIFitis** [comment]
@skatrak Thanks for the review. I've adrressed the latest comments in the new revision.

**Comment by TIFitis** [comment]
Ping for review.

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (skatrak) [inline flang/lib/Lower/OpenMP/OpenMP.cpp:697]: Nit: Add a description to the assert.

**Review by skatrak — APPROVED** [review]
Thank you, LGTM!

**Pull request merged**
